# syntax-transform/3.1/0018-regexp-encoding.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, Ruby does not throw an error when performing a regular expression
# operation on a string that is incorrectly encoded or if the regular expression
# and the string's encodings are incompatible. Also, in Ruby versions later than
# 1.8, the default encoding for a regular expression if not specified using the
# `e`, `n`, `s` or `u` regular expression parameters is 7-bit ASCII if the
# regular expression can be encoded in 7-bit ASCII, or the current script
# encoding otherwise, whereas in Ruby 1.8, the default encoding for a regular
# expression if not specified is UTF-8 (it should actually be 8-bit ASCII in
# Ruby 1.8, but it can be configured at compile time using the
# `--with-default-kcode` configure flag, and RPG Maker XP and RPG Maker VX
# appear to have configured the default regular expression encoding to be
# UTF-8).
#
# Thus, this code should print `7` and `[3, 3, 1]` in Ruby 1.8 if the default
# regular expression encoding is correctly configured to be UTF-8 like in RPG
# Maker XP and RPG Maker VX. In later versions, it prints `7` and
# `[1, 1, 1, 1, 1, 1, 1]`.
#
#   a = [0xe5, 0xa4, 0xa9, 0xe5, 0xe5, 0x9c, 0xb0].pack('c*')
#   p a.length
#   p a.scan(/./m).map(&:length)

--- a/enc/utf_8.c
+++ b/enc/utf_8.c
@@ -228,6 +228,9 @@ mbc_enc_len(const UChar* p, const UChar* e, OnigEncoding enc ARG_UNUSED)
 {
   int firstbyte = *p++;
   state_t s;
+  if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) { /* The length of a UTF-8 multi-byte character is based only on the first byte of the character in Ruby 1.8 */
+    return (firstbyte & 0x80) == 0 ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(1) : (firstbyte & 0xe0) == 0xc0 ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(2) : (firstbyte & 0xf0) == 0xe0 ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(3) : (firstbyte & 0xf8) == 0xf0 ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(4) : (firstbyte & 0xfc) == 0xf8 ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(5) : (firstbyte & 0xfe) == 0xfc ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(6) : ONIGENC_CONSTRUCT_MBCLEN_INVALID();
+  }
   s = trans[0][firstbyte];
   if (s < 0) return s == ACCEPT ? ONIGENC_CONSTRUCT_MBCLEN_CHARFOUND(1) :
                                   ONIGENC_CONSTRUCT_MBCLEN_INVALID();
--- a/re.c
+++ b/re.c
@@ -1503,6 +1503,12 @@ rb_reg_prepare_enc(VALUE re, VALUE str, int warn)
     rb_encoding *enc = 0;
     int cr = str_coderange(str);
 
+    /* If syntax transform is active and targeting Ruby <= 1.8, ignore encoding errors in strings when performing regular expression operations on them and make the encoding of a regular expression default to UTF-8 if not specified */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        rb_reg_check(re);
+	return rb_reg_fixed_encoding_p(re) ? RREGEXP_PTR(re)->enc : rb_utf8_encoding();
+    }
+
     if (cr == ENC_CODERANGE_BROKEN) {
         rb_raise(rb_eArgError,
             "invalid byte sequence in %s",
